networktest.go

```diff
@@ -235,87 +235,110 @@ func (l *lightningNode) Start(lndError chan error) error {
 	// process errors to the goroutine running the tests.
 	go func() {
 		err := l.cmd.Wait()
+		if err != nil {
+			lndError <- errors.Errorf("%v\n%v\n", err, errb.String())
+		}
 
 		// Signal any onlookers that this process has exited.
 		close(l.processExit)
-
-		if err != nil {
-			lndError <- errors.New(errb.String())
-		}
 	}()
 
-	pid, err := os.Create(filepath.Join(l.cfg.DataDir,
-		fmt.Sprintf("%v.pid", l.nodeID)))
+	// Write process ID to a file.
+	if err := l.writePidFile(); err != nil {
+		l.cmd.Process.Kill()
+		return err
+	}
+
+	// Since Stop uses the LightningClient to stop the node, if we fail to get a
+	// connected client, we have to kill the process.
+	conn, err := l.connectRPC()
 	if err != nil {
+		l.cmd.Process.Kill()
 		return err
 	}
-	l.pidFile = pid.Name()
-	if _, err = fmt.Fprintf(pid, "%v\n", l.cmd.Process.Pid); err != nil {
+	l.LightningClient = lnrpc.NewLightningClient(conn)
+
+	// Obtain the lnid of this node for quick identification purposes.
+	ctxb := context.Background()
+	info, err := l.GetInfo(ctxb, &lnrpc.GetInfoRequest{})
+	if err != nil {
+		return err
+	}
+
+	l.PubKeyStr = info.IdentityPubkey
+
+	pubkey, err := hex.DecodeString(info.IdentityPubkey)
+	if err != nil {
 		return err
 	}
-	if err := pid.Close(); err != nil {
+	copy(l.PubKey[:], pubkey)
+
+	// Launch the watcher that'll hook into graph related topology change
+	// from the PoV of this node.
+	l.wg.Add(1)
+	go l.lightningNetworkWatcher()
+
+	return nil
+}
+
+// writePidFile writes the process ID of the running lnd process to a .pid file.
+func (l *lightningNode) writePidFile() error {
+	filePath := filepath.Join(l.cfg.DataDir, fmt.Sprintf("%v.pid", l.nodeID))
+
+	pid, err := os.Create(filePath)
+	if err != nil {
 		return err
 	}
 
+	_, writeErr := fmt.Fprintf(pid, "%v\n", l.cmd.Process.Pid)
+	closeErr := pid.Close()
```

Comment:
Is it possible to defer this `Close()` after creating the file?

Reply:
Sure. The downside to doing it with `defer` though is that if `Close` returns an error, we won't be able to return that error from this function.